File: Algorithms/230. Kth Smallest Element in a BST/Solution.py
#coding=utf-8
__author__ = 'xuxuan'
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def kthSmallest(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: int
        """
        return self.f(root)[k-1]
    def f(self,root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        return [] if root is None else self.f(root.left)+[root.val]+self.f(root.right)

File: Algorithms/230. Kth Smallest Element in a BST/Solution.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @param {Integer} k
# @return {Integer}
def kth_smallest(root, k)
    leftNum=calTreeNum(root.left)
    if leftNum+1==k
        return root.val
    end
    return leftNum+1>k ? kth_smallest(root.left, k) : kth_smallest(root.right, k-leftNum-1)
end

# @param {TreeNode} root
# @return {Integer}
def calTreeNum(root)
    return root ? 1+calTreeNum(root.left)+calTreeNum(root.right) : 0
end